plan_monitor/collect.py

```diff
@@ -0,0 +1,154 @@
+                logger.debug(f'Poll retrieved {count:,} rows in {q_time:,} ms.')
+    except KeyboardInterrupt:
+        logger.info('Exiting due to external interrupt request.')
+        exit_message_logged = True
+    except Exception:
+        logger.info('Exiting due to unhandled exception.')
+        exit_message_logged = True
+        raise
+    finally:
+        conn.close()
+        result_queue.close()
+        stop_event.set()
+        if not exit_message_logged:
+            logger.info('Exiting due to shutdown initiated by another subprocess.')
+
+
+def collect() -> None:
+    schema_registry = confluent_kafka.schema_registry.SchemaRegistryClient({'url': config.SCHEMA_REGISTRY_URL})
+    key_serializer = confluent_kafka.schema_registry.avro.AvroSerializer(
+        message_schemas.STATS_MESSAGE_KEY_AVRO_SCHEMA, schema_registry)
+    value_serializer = confluent_kafka.schema_registry.avro.AvroSerializer(
+        message_schemas.STATS_MESSAGE_VALUE_AVRO_SCHEMA, schema_registry)
+    producer_config = {'bootstrap.servers': config.KAFKA_BOOTSTRAP_SERVERS,
+                       'key.serializer': key_serializer,
+                       'value.serializer': value_serializer,
+                       'linger.ms': 100,
+                       'retry.backoff.ms': 250,
+                       'compression.codec': 'snappy'}
+    kafka_producer = confluent_kafka.SerializingProducer(producer_config)
+    result_queue = mp.Queue(10000)
+    stop_event = mp.Event()
+    produced_count = 0
+    processes = []
+
+    for db_identifier, odbc_conn_string in config.ODBC_CONN_STRINGS.items():
+        processes.append(mp.Process(target=poll_db, name=f'db-{db_identifier}',
+                                    args=(db_identifier, odbc_conn_string, stop_event, result_queue)))
+
+    start_time = time.perf_counter()
+
+    for process in processes:
+        process.start()
+        # Stagger so the herd doesn't dump a ton onto the results queue or Kafka producer at once:
+        time.sleep(0.5)
+
+    logger.info(f'Started {len(processes)} DB poll subprocesses')
+
+    try:
+        while not (stop_event.is_set() and result_queue.empty()):
+            kafka_producer.poll(0)  # serve delivery callbacks
+            try:
+                msg_value = result_queue.get(timeout=0.1)
+            except queue.Empty:
+                continue
+            msg_key = {"db_identifier": msg_value["db_identifier"],
+                       "set_options": msg_value["set_options"],
+                       "sql_handle": msg_value["sql_handle"]}
+            kafka_producer.produce(topic=config.STATS_TOPIC, key=msg_key, value=msg_value,
+                                   on_delivery=common.kafka_producer_delivery_cb)
+            produced_count += 1
+            if produced_count % 100_000 == 0:
+                logger.info(f"Produced {produced_count:,} stats records since process start...")
+                if not all([p.is_alive() for p in processes]):
```

Comment:
is this basically an uncaught exception?

Reply:
In prior iterations I thought I had seen subprocesses "hanging" even after they had thrown an exception. In the most recent code I could not reproduce that, so I think I must have changed something (probably around the handling of the queue) that fixed that issue. As such, this is now removed in https://github.com/woodlee/sqlserver-plan-regression-monitor/pull/1/commits/3556ebf3edd292a41c46c2cbe06e30a2d70b9002. That commit does a little further cleanup around multiprocessing stuff too, so please have a look!